### PR TITLE
pkg/{otlp,trace}: introduce APM stats consumer

### PR DESCRIPTION
**Pull request by gbbr**
This change introduces an APM Stats consumer into the OTLP Ingest pipeline to ensure that APM Stats coming in as OTLP Metrics generated by the Datadog Processor in the OpenTelemetry Collector correctly get flushed.

The logic for converting between APM Stats and Metrics that is used by both the OTLP Ingest and datadogprocessor lives in the shared pkg/otlp/model package.

### CODE REVIEW

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:61]: Does this need to be public?

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:99]: Instead of prefixing logs with `StatsPayload Translator`, add a new attribute referring to this (you could even create a new logger from `t.logger` that always appends this field using the [`With`](https://pkg.go.dev/go.uber.org/zap#Logger.With) method)

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:175]: I think this should be fine, but if we want to be fully accurate, we can change the mapping on the ddsketch itself instead of doing this.

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:312]: Does this need to be public?

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:319]: Don't use formatted strings with the zap logger, use fields instead for performance

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:363]: Do we choose these names? If so, I think we should namespace them (`dd.stats.hits`?) to avoid collisions

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [inline pkg/otlp/model/translator/statspayload.go:381]: If any of these assumptions do not hold, something has gone really wrong. I think it would make our life easier if we fail more loudly (say, by returning the error and dropping the whole payload), since we will be able to detect these errors more easily in the wild

File: pkg/otlp/model/translator/metrics_translator_test.go
Comment (mx-psi) [inline pkg/otlp/model/translator/metrics_translator_test.go:587]: Adding a comment so that we don't forget to solve/discuss this before merging

File: pkg/otlp/model/translator/sketches_test.go
Comment (mx-psi) [inline pkg/otlp/model/translator/sketches_test.go:39]: Adding a comment so that we don't forget to solve/discuss this before merging

File: pkg/otlp/model/translator/testhelper_test.go
Comment (mx-psi) [inline pkg/otlp/model/translator/testhelper_test.go:122]: Adding a comment so that we don't forget to solve/discuss this before merging

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (gbbr) [reply pkg/otlp/internal/serializerexporter/consumer.go:127]: I generally think it's a better approach to mark a struct's API by making its methods exported. It makes it more obvious how it is used. It seems unnatural to call unexported methods from outside a struct's own methods. This struct is private, nobody can use it outside this package. The public methods simply signify its usage. If everything is unexported, it's hard to figure that out. What do you think? Does it make sense?

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (gbbr) [reply pkg/otlp/internal/serializerexporter/consumer.go:150]: The datadogprocessor will compute stats of all the traces that pass through it. If those traces don't reach our intake because of configuration or (fully exclusive) sampling, then yes, there could be (valid) stats without traces.

As for traces without stats. I don't think that can happen as long as the processor is used. If the processor is not used, then we are where we are today: stats only for sampled traces.

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (gbbr) [reply pkg/otlp/internal/serializerexporter/consumer.go:155]: Agreed, that's better.

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (gbbr) [reply pkg/otlp/internal/serializerexporter/consumer.go:162]: Done.

File: pkg/otlp/model/translator/metrics_translator.go
Comment (gbbr) [reply pkg/otlp/model/translator/metrics_translator.go:445]: Probably not in most cases. But in a gateway setup where things could happen in between the processor and the translator, who knows? Maybe yes. Either way, I'm changing this in my next commit.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go:61]: Yes. It is used in the processor.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go:175]: We want the ddsketch to be identical and consistent with what the trace-agent / exporter generates in all other cases. The ddsketch is the important piece of truth. It must not be changed. The ExponentialHistogram is just an intermediate structure temporarily holding the data. It should never be used. If it is ever used, it won't be with Datadog and in that case a little inaccuracy does not matter.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go:312]: This one not so much, but I'd keep it for symmetry. Is that a bad reason?

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go:319]: That would put the field at the end as JSON, making it a pretty ugly log message. I could inline the text, or define the message globally, but what performance are we talking about here? This is generally called once every 10 seconds, and this log line should in reality never happen.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go:31]: Agreed. But I also don't think that's important now. I remember trying to do that in the past and it ended up being complicated because that package depends on other packages in `pkg/trace`. So far, everything using `pkg/otlp/model` also uses `pkg/trace` so we should be good.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go:99]: I created a child logger with a `component: metrics translator` field and toned down on the prefixes.

**Review by gbbr — COMMENTED** [review]
Thanks for the review. I addressed all your comments except [this one](https://github.com/DataDog/datadog-agent/pull/14586/files#r1039446281), which I'm addressing in the next commit since it is a bigger change.

File: pkg/otlp/model/translator/metrics_translator.go
Comment (gbbr) [reply pkg/otlp/model/translator/metrics_translator.go]: Done. I've changed the code so that any set of resource metrics with that attribute would get processed, not just the first!

**Review by gbbr — COMMENTED** [review]
Addressed second comment too. Should be good for another review.

**Review by buraizu — APPROVED** [review]

**Review by dianashevchenko — APPROVED** [review]
Regarding `pkg/trace/{api/api.go,agent/agent.go}` : lgtm

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [reply pkg/otlp/model/translator/statspayload.go:31]: Will write more about this in an upcoming proposal, let's mark it as solved for now :)

File: pkg/otlp/model/translator/metrics_translator.go
Comment (mx-psi) [reply pkg/otlp/model/translator/metrics_translator.go]: Awesome, I think that's a more robust approach

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [reply pkg/otlp/model/translator/statspayload.go]: If someone wants to make this public in the future, that's easy, if we want to make it private it's (potentially) harder because of any dependencies we may have introduced, so I think in principle it's good to limit our public API as much as possible. I don't want to block the PR on this, so I will leave it up to you to decide since I don't think it's very bad if it remains public.

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [reply pkg/otlp/model/translator/statspayload.go]: > That would put the field at the end as JSON, making it a pretty ugly log message

Formatting will depend on where this is used and how you have configured the logger, in the Agent this would not be JSON, in the Collector it will be by default but you can configure it to have it look as you wish (and it's very common there to have fields on logging messages)

The performance impact will be small in this case, to me this is more of a 'this is not how zap's authors recommend using this API' and 'setting a good example in our codebase' than anything else.

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (mx-psi) [reply pkg/otlp/internal/serializerexporter/consumer.go:125]: That makes sense, yes, it won't make much of a difference and I agree it's relevant to mark the struct's API

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (mx-psi) [reply pkg/otlp/internal/serializerexporter/consumer.go:149]: I was thinking about the failure mode where sending stats to `/v0.6/stats` would fail but sending traces to the `5003` port would not because something is set up incorrectly in the configuration (say some configuration flag that only affects one or the other). This seems pretty straightforward so probably there is no such case, but wanted to double check with you.

File: pkg/otlp/model/translator/statspayload.go
Comment (mx-psi) [reply pkg/otlp/model/translator/statspayload.go:185]: The difference seems negligible so I guess we can ignore this for now and revisit if it becomes a problem

**Review by mx-psi — APPROVED** [review]
The PR LGTM. I left some non-blocking questions and some comments that would be nice to address:

- https://github.com/DataDog/datadog-agent/pull/14586#discussion_r1042028357
- https://github.com/DataDog/datadog-agent/pull/14586#discussion_r1042016410
- https://github.com/DataDog/datadog-agent/pull/14586#discussion_r1042019469

None of the comments are very concerning to me, so I will defer to your judgement on what the ultimate decision is on those.

File: pkg/otlp/internal/serializerexporter/consumer.go
Comment (gbbr) [reply pkg/otlp/internal/serializerexporter/consumer.go:149]: I'm not aware of any such case. The receiver port needs to be correct and active for traces to come in. I don't think that's ever disabled (except when running the agent in the collector exporter, but that's not the case here since we are in the Datadog Agent and the trace-agent receiver is on by default).

As for the 5003 internal port, I don't see why users would want to change that, unless it clashes with something. But then again, if traces don't work, it's usually a higher priority than stats so it should be noticed quicker.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go]: I've already made this private in my latest commit with the tests. Forgot to follow up here.

File: pkg/otlp/model/translator/statspayload.go
Comment (gbbr) [reply pkg/otlp/model/translator/statspayload.go]: I agree with what you said, and you may have noticed I've used fields wherever it made sense, but here I think it doesn't. This is the current message:
```
was asked to convert metrics to stats payload, but identifier key "_dd.apm_stats" was not present. Skipping.
```
If I were to use a field, it would be:
```
was asked to convert metrics to stats payload, but identifier key was not present. Skipping {"key": "_dd.apm_stats"}
```
The latter seems very unusual to me. Why would you do that? I see fields as context to the message, am I wrong? I don't think that's the case here.

**Pull request merged**
